tests/unit/services/popup-test.js

```diff
@@ -1,180 +1,237 @@
-      assert.ok(false, 'resolved the open promise');
-    }, function(){
-      assert.ok(true, 'rejected the open promise');
-    }).finally(done);
+  test('open rejects when window does not open', function (assert) {
+    let done = assert.async();
+
+    const closedWindow = buildMockWindow();
+
+    closedWindow.closed = true;
+    window.open = function () {
+      assert.ok(true, 'calls window.open');
+      return closedWindow;
+    };
+
+    run(function () {
+      popup
+        .open('http://some-url.com', ['code'])
+        .then(
+          function () {
+            assert.ok(false, 'resolves promise');
+          },
+          function () {
+            assert.ok(true, 'rejected the open promise');
+            assert.ok(
+              !localStorage.getItem(CURRENT_REQUEST_KEY),
+              'current request key is removed'
+            );
+          }
+        )
+        .finally(done);
+    });
   });
 
-  mockWindow.closed = true;
-});
+  test('open does not resolve when receiving a storage event for the wrong popup id', function (assert) {
+    let done = assert.async();
+
+    window.open = function () {
+      assert.ok(true, 'calls window.open');
+      return buildMockWindow();
+    };
+
+    const promise = run(function () {
+      return popup
+        .open('http://someserver.com', ['code'])
+        .then(
+          function () {
+            assert.ok(false, 'resolves the open promise');
+          },
+          function () {
+            assert.ok(false, 'rejected the open promise');
+          }
+        )
+        .finally(done);
+    });
 
-test("localStorage state is cleaned up when parent window closes", function(assert){
-  var mockWindow = buildMockWindow();
-  window.open = function(){
-    assert.ok(true, 'calls window.open');
-    return mockWindow;
-  };
+    localStorage.setItem(
+      PopupIdSerializer.serialize('invalid'),
+      'http://authServer'
+    );
+    // Need to manually trigger storage event, since it doesn't fire in the current window
+    window.dispatchEvent(buildMockStorageEvent('invalid', 'http://authServer'));
+
+    setTimeout(function () {
+      assert.ok(
+        !promise.isFulfilled,
+        'promise is not fulfulled by invalid data'
+      );
+      assert.deepEqual(
+        'http://authServer',
+        localStorage.getItem(PopupIdSerializer.serialize('invalid')),
+        "doesn't remove the key from local storage"
+      );
+      done();
+    }, 10);
+  });
 
-  run(function(){
-    popup.open('some-url', ['code']).then(function(){
-      assert.ok(false, 'resolved the open promise');
-    }, function(){
-      assert.ok(false, 'rejected the open promise');
+  test('open rejects when window closes', function (assert) {
+    let done = assert.async();
+
+    const mockWindow = buildMockWindow();
+    window.open = function () {
+      assert.ok(true, 'calls window.open');
+      return mockWindow;
+    };
+
+    run(function () {
```

Comment:
Looking at my own branch, I think these `run()` wrappers are not needed anymore?

Reply:
You are probably correct. I'll try it without them.